tutorials/04 - DistortionEffectDemo.md

```diff
@@ -0,0 +1,138 @@
+```
+if (! isEnabled())
+    return;
+```
+
+- We iterate through each sample of each channel and apply the tanh waveshaping function to the output audio.
+```
+for (int channel = 0; channel < fc.destBuffer->getNumChannels(); ++channel)
+{
+    auto dest = fc.destBuffer->getWritePointer (channel);
+                
+    for (int i = 0; i < fc.bufferNumSamples; ++i)
+        dest[i] = std::tanh (gainValue * dest[i]);
+}
+```
+
+## DistortionEffectDemo Class
+
+The DistortionEffectDemo is defined similarly to the PitchAndTimeComponent class
+
+```
+class DistortionEffectDemo : public Component,
+                             private ChangeListener
+```
+
+## Private Members
+- If you look in the private member section of the `DistortionEffectDemo` again, you'll see similarities to the PitchAndTimeComponent and PlayBackDemo. There is an `Engine` to set up Tracktion Engine, and an `Edit` for editing.
+
+```
+te::Edit edit { engine, te::createEmptyEdit(), te::Edit::forEditing, nullptr, 0 };
+te::TransportControl& transport { edit.getTransport() };
+```
+- We've added a slider called `gainSlider` to represent distortion gain.
+```
+TextButton settingsButton { "Settings" }, playPauseButton { "Play" }, loadFileButton { "Load file" };
+Thumbnail thumbnail { transport };
+Slider gainSlider;
+```
+
+### DistortionEffectDemo() Constructor
+
+- In the constructor, we'll start by registering the `DistortionPlugin` class with Tracktion Engine.
+```
+engine.getPluginManager().createBuiltInType<DistortionPlugin>();
+```
+
+- We'll add the sliders and buttons to the GUI.
+```
+Helpers::addAndMakeVisible (*this, { &gainSlider, &settingsButton, &playPauseButton });
+```
+
+- Next, we'll load the PlayBackDemo audio example into file f.
+```
+oggTempFile = std::make_unique<TemporaryFile> (".ogg");
+auto f = oggTempFile->getFile();
+f.replaceWithData (PlaybackDemoAudio::BITs_Export_2_ogg, PlaybackDemoAudio::BITs_Export_2_oggSize);
+```
+
+- We'll create a clip from the audio file. Then we'll create a track on the edit. Finally, we'll load the clip onto the track.
+```
+auto track = EngineHelpers::getOrInsertAudioTrackAt (edit, 0);
+jassert (track != nullptr);
+
+te::AudioFile audioFile (edit.engine, f);
+auto clip = track->insertWaveClip (f.getFileNameWithoutExtension(), f,
+                                  { { 0.0, audioFile.getLength() }, 0.0 }, false);
+jassert (clip != nullptr);
+```
+
+- Now that the track has been created and the clip is on the track, we need to create a new instance of DistortionPlugin and load it onto the track.
+```
+auto gainPlugin = edit.getPluginCache().createNewPlugin (DistortionPlugin::xmlTypeName, {});
+track->pluginList.insertPlugin (gainPlugin, 0, nullptr);
+```
+- Finally, in order to hear the effect being played, we must set the transport loop range to the start/end of the clip, set it to loop, and initialise play.
+```
+edit.getTransport().addChangeListener (this);
+EngineHelpers::loopAroundClip (*clip);
+```
+
```

Comment:
Before you write this I'd add a section on the `ParameterValueSource` class explaining what it does and why it needs to listen to the AutomatableParameter. You can use the two sliders as an example so they stay in sync. In a real DAW you can have multiple representations of a parameter e.g. a curve (which will just be straight line if it has no points), a rotary slider in the properties panel parameter list, a quick control slider on the plugin mixer UI element and so on. Basically this is a quick, concise way of attaching Value backed Components to AutomatableParameters.